keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,914 @@
+  if opts.AlignByUnCoreCacheOption && opts.DistributeCPUsAcrossCores {
+    return opts, fmt.Errorf("static policy options %s and %s can not be used at the same time", AlignByUnCoreCacheOption, DistributeCPUsAcrossCoresOption)
+  }
+  ```
+
+- In `static_policy.go`, check for option `AlignByUnCoreCacheOption` in takeByTopology. Implement new cpu assignment function that takeByTopologyUncorecachePacked which adds a third level of CPU grouping in the topology. **Socket->Numa->UnCoreCache**. TakeByTopologyUnCoreCachePacked uses the "packed" cpu sorting strategy similar to takeByTopologyNUMAPacked.
+
+  ```go
+  func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int) (cpuset.CPUSet, error) {
+  ...
+    acc.numaOrSocketsFirst.takeFullFirstLevel()
+    if acc.isSatisfied() {
+     return acc.result, nil
+     }
+    acc.numaOrSocketsFirst.takeFullSecondLevel()
+    if acc.isSatisfied() {
+      return acc.result, nil
+    }
+   // 2. Acquire partial uncorecache, if there are enough CPUs available to satisfy the container requirement
+   //    Acquire the full uncorecache, if available and the container requires at least all the CPUs in the uncorecache grouping
+    acc.numaOrSocketsFirst.takeThirdLevel()
+    if acc.isSatisfied() {
+        return acc.result, nil
+    }
+  ...
+  }
+  ```
+
+  - Define a new interface in the `numaOrSocketsFirstFunc` in `cpu_assignment.go` that implements `takeThirdLevel`, which calls `takeUnCoreCache`.
+
+  ```go
+  // In Split L3 Topology, we take from the sets of uncorecache as the third level
+  func (n *numaFirst) takeThirdLevel() {
+    n.acc.takeUnCoreCache()
+  }
+  // First try to take partial uncorecache (CCD), if available and the request size can fit w/in the uncorecache.
+  // Second try to take the full CCD if available and need is at least the size of the uncorecache group.
+  func (a *cpuAccumulator) takeUnCoreCache() {
+    // check if SMT ON
+    
+    for _, uncore := range a.allUnCoreCache() {
+      numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() //this is another new change
+      
+      var freeCPUsInUncorecache cpuset.CPUSet
+      //need to get needed cores in uncorecache
+      freeCoresInUncorecache := a.details.CoresNeededInUnCoreCache(numCoresNeeded, uncore)
+      klog.V(2).Info("free cores from a.details list: ", freeCoresInUncorecache)
+      for _, coreId := range freeCoresInUncorecache.List() {
+          freeCPUsInUncorecache = freeCPUsInUncorecache.Union(a.topo.CPUDetails.CPUsInCores(coreId))
+      }
+      klog.V(2).Info("freeCPUsInUncorecache  : ", freeCPUsInUncorecache)
+      if a.numCPUsNeeded == freeCPUsInUncorecache.Size() {
+          klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", uncore)
+        a.take(freeCPUsInUncorecache)
+      }
+      //take full Uncorecache if the numCPUsNeeded is greater the L3 cache size
+      a.takeFullUnCore
+      if a.isSatisfied() {
+        return
+      }
```

Comment:
reprising from https://github.com/kubernetes/enhancements/pull/4810/files#r1750398234 this seems indeed to take partial first and then full, which is different from the overall logic. Can we elaborate the reasoning for this diversion, if we have a diversion at all (please keep me honest)